docs/js-dev-standards.md

```diff
@@ -0,0 +1,101 @@
+# JavaScript Software Development Standards
+
+The goal of this document is to describe how our team’s projects are typically structured. This includes normative tooling and structure for different kinds of software (e.g. libraries vs applications), and what typical flows a developer should expect to be available.
+
+
+## Goals and Principles
+* __Consist__ workflow across projects
+* __Familiar__ to developers, and easy to adopt
+* __Fast__ CI/CD and development/testing workflows. To support this principle, we transpile and bundle code _as little as possible_ and support dev-loops that only use source code.
+* __Modern__ We lean into the latest features of the JavaScript ecosystem, including TypeScript, ESM, and modern language features.
+
+## Project Structure
+Our projects are structured as monorepos using Yarn 3+ (or whatever is latest). Our software is usually composed of many moving parts: libraries, demo applications, full-blown applications, test suites, and more. To manage these in a single repository, we use Yarn workspaces.
+
+### JavaScript-Only Monorepos
+For single-language monorepos, workspaces are created under the `packages` folder. For example, a monorepo with a library and a demo application would look like this:
+
+```text
+- packages
+	- my-library
+		- package.json
+	- my-demo
+		- package.json
+```
+
+### Multi-Language Monorepos
+When authoring monorepos with multiple languages, Yarn workspaces are created under the `javascript` folder, and other languages would follow a similar pattern. For example, a monorepo with a library, demo application, and a Go backend would look like this:
+
+```text
+- javascript
+  - my-library
+		- package.json
+	- my-demo
+		- package.json
+- go
+	- my-backend
+		- go.mod
+		- go.sum
+		- main.go
+```
+
+The rest of this document will focus primarily on JavaScript.
+
+## Common Build Tasks
+The following top-level tasks expected to exist in all JavaScript projects. 
+
+* __clean__ - this task cleans the project, removing any artifacts or temporary files.
+* __release__ - this task applies semver changes, executes build verification, and publishes artifacts to the package registry.
+* __ci__ - this task is used to chain together an end-to-end continuous verification pipeline. It should run all tests, linting, and other verification tasks.
+* __start__ - this task starts any webapps or servers in the project in development mode.
+* __build__ - this task builds artifacts and performs typechecking for library packages within the project.
+* __test__ - this task runs all tests in the project.
+* __check__ - this task checks code for linting and formatting issues.
+* __fix__ - this task checks code for linting and formatting issues and attempts to fix as many as possible.
+
+
+## CI Tasks
```

Comment:
So in some projects we favor CI/CD readability (e.g. split lint, test, build, etc..), but this comes at a cost in terms of performance. `turbo` can't schedule these tasks optimally in this case.

Moreover, if we move to the fully-parallel versions of check, test, etc.. that use `yarn workspaces foreach`, I think we'll be making a steeper performance regression.

Reply:
I've seen some projects create files for every type of check (lint, test, formatting ,etc..), but these get super hairy. T

he amount of time you can actually save by partitioning actions is highly subjective. In most of our library-monorepos, the CI script takes 2-5 minutes, which is at the point where it's not worth complicating with parallelism.

Reply:
So right now we primarily just split out version-check since that is very fast and doesn't depend on turbo. Does it make sense to do a fast-fail of something simple like a lint check before committing to a full build cycle?

Reply:
We should be able to get that by reordering the ci task deps: 

e.g. `"ci": ( "dependsOn": ["lint", "build", "test", "bundle"] }`

The fail-fast stuff I mentioned earlier has to do with how the lint runs - if it's running in 'fix' mode then the CI will fail on `git_is_clean` , which is undesirable. This is why we're recommending `check` and `fix` as replacements for `lint`

Reply:
So we could split it out into a separate github action, and it would be very clear that the lint failed if it were its own separate action, but the complexity cost couuld get pretty high. And we'd lose the benefit of a lot of our monorepo orchestration work and turbo-caching in GH actions.